Light Vox Engine/FrameResource.cpp

```diff
@@ -0,0 +1,155 @@
+	cbvSrvGpuHandle.Offset(nullSrvCount + textureCount + (frameResourceIndex * LV_FRAME_COUNT), cbvSrvDescriptorSize);
+
+	// Create the constant buffers.
+	const UINT constantBufferSize = (sizeof(SceneConstantBuffer) + (D3D12_CONSTANT_BUFFER_DATA_PLACEMENT_ALIGNMENT - 1)) & ~(D3D12_CONSTANT_BUFFER_DATA_PLACEMENT_ALIGNMENT - 1); // must be a multiple 256 bytes
+	ThrowIfFailed(device->CreateCommittedResource(
+		&CD3DX12_HEAP_PROPERTIES(D3D12_HEAP_TYPE_UPLOAD),
+		D3D12_HEAP_FLAG_NONE,
+		&CD3DX12_RESOURCE_DESC::Buffer(constantBufferSize),
+		D3D12_RESOURCE_STATE_GENERIC_READ,
+		nullptr,
+		IID_PPV_ARGS(&sceneConstantBuffer)));
+
+	// Map the constant buffers and cache their heap pointers.
+	CD3DX12_RANGE readRange(0, 0);		// We do not intend to read from this resource on the CPU.
+	ThrowIfFailed(sceneConstantBuffer->Map(0, &readRange, reinterpret_cast<void**>(&sceneConstantBufferWO)));
+
+	// Create the constant buffer view for scene pass
+	D3D12_CONSTANT_BUFFER_VIEW_DESC cbvDesc = {};
+	cbvDesc.SizeInBytes = constantBufferSize;
+
+	// Describe and create the scene constant buffer view (CBV) and 
+	// cache the GPU descriptor handle.
+	cbvDesc.BufferLocation = sceneConstantBuffer->GetGPUVirtualAddress();
+	device->CreateConstantBufferView(&cbvDesc, cbvSrvCpuHandle);
+	sceneCbvHandle = cbvSrvGpuHandle;
+
+	//combine all of our command lists
+	const UINT batchSize = _countof(sceneCommandLists) + LV_COMMAND_LIST_COUNT; // _countof(m_shadowCommandLists) + 3;
+	batchedCommandList[0] = commandLists[LV_COMMAND_LIST_PRE].Get();
+	memcpy(batchedCommandList + 1, sceneCommandLists, _countof(sceneCommandLists) * sizeof(ID3D12CommandList*));
+	batchedCommandList[_countof(sceneCommandLists) + 1] = commandLists[LV_COMMAND_LIST_MID].Get();
+	batchedCommandList[_countof(sceneCommandLists) + 2] = commandLists[LV_COMMAND_LIST_POST].Get();
+}
+
+FrameResource::~FrameResource()
+{
+	for (int i = 0; i < LV_COMMAND_LIST_COUNT; i++)
+	{
+		commandAllocators[i] = nullptr;
+		commandLists[i] = nullptr;
+	}
+
+	sceneConstantBuffer = nullptr;
+
+	for (int i = 0; i < LV_NUM_CONTEXTS; i++)
+	{
+		sceneCommandLists[i] = nullptr;
+		sceneCommandAllocators[i] = nullptr;
+	}
+}
+
+void FrameResource::Bind(ID3D12GraphicsCommandList * commandList, BOOL scenePass, D3D12_CPU_DESCRIPTOR_HANDLE * rtvHandle, D3D12_CPU_DESCRIPTOR_HANDLE * dsvHandle)
+{
+	//set-up before rendering
+	commandList->SetGraphicsRootDescriptorTable(1, sceneCbvHandle);
+	commandList->OMSetRenderTargets(1, rtvHandle, FALSE, dsvHandle);
+}
+
+void FrameResource::Init()
+{
+	// Reset the command allocators and lists for the main thread.
+	for (int i = 0; i < LV_COMMAND_LIST_COUNT; i++)
+	{
+		ThrowIfFailed(commandAllocators[i]->Reset());
+		ThrowIfFailed(commandLists[i]->Reset(commandAllocators[i].Get(), pso.Get()));
+	}
+
+	// Reset the worker command allocators and lists.
+	for (int i = 0; i < LV_NUM_CONTEXTS; i++)
```

Comment:
would it be best to change these 2 to something like
`for(auto i = 0; i < LV_COMMAND_LIST_COUNT; ++i)` ?